### PR TITLE
feat: send flag decisions

### PR DESCRIPTION
**Pull request by oakbani**
## Summary
- Dispatches event when rollout and send_flag_deciions flag is True
- Dispatches event when variation is nil ( neither feature-test nor rollout) and send_flag_decisions is True
- Adds metadata to impression event. 

## Test plan
- Fixed All Unit tests
- Verified Manually with FSC Flag Decisions Tests

### CODE REVIEW

File: lib/optimizely.rb [hunk 1]
```diff
@@ -140,7 +140,10 @@ def activate(experiment_key, user_id, attributes = nil)
 
       # Create and dispatch impression event
       experiment = config.get_experiment_from_key(experiment_key)
-      send_impression(config, experiment, variation_key, user_id, attributes)
+      send_impression(
+        config, experiment, variation_key, experiment_key, experiment_key,
```
Comment (pawels-optimizely) [inline lib/optimizely.rb:144]: why are we passing experiment key twice ?

File: spec/project_spec.rb [hunk 2]
```diff
@@ -154,7 +154,13 @@ class InvalidErrorHandler; end
             decisions: [{
               campaign_id: '1',
               experiment_id: '111127',
-              variation_id: '111128'
+              variation_id: '111128',
+              metadata: {
+                flag_key: 'test_experiment',
```
Comment (pawels-optimizely) [inline spec/project_spec.rb:159]: this is never possible when flag_key is rule_key

File: spec/project_spec.rb [hunk 3]
```diff
@@ -222,6 +228,13 @@ class InvalidErrorHandler; end
       }]
       params[:visitors][0][:snapshots][0][:events][0][:entity_id] = '3'
 
+      params[:visitors][0][:snapshots][0][:decisions][0][:metadata] = {
+        flag_key: 'test_experiment_with_audience',
```
Comment (pawels-optimizely) [inline spec/project_spec.rb:232]: same comment as before

**Review by pawels-optimizely — CHANGES_REQUESTED** [review]

**Comment by oakbani** [comment]
@pawels-optimizely  As discussed, 
- now activate will pass empty string for flag key
- is_feature_enabled will pass empty string for rule key and variation key and rollout in rule type when variation is nil.

**Review by msohailhussain — APPROVED** [review]
lgtm

**Review by pawels-optimizely — APPROVED** [review]
lgtm

**Review by pawels-optimizely — CHANGES_REQUESTED** [review]
please address my comments

**Review by pawels-optimizely — APPROVED** [review]
lgtm, thx for addressing the comments

**Pull request merged**